server/src/main/java/org/apache/druid/server/coordination/ServerManager.java

```diff
@@ -204,22 +224,45 @@ public ServerManager(
         analysis.getBaseQuery().orElse(query)
     );
 
-    final FunctionalIterable<QueryRunner<T>> queryRunners = FunctionalIterable
-        .create(specs)
-        .transformCat(
-            descriptor -> Collections.singletonList(
-                buildQueryRunnerForSegment(
+    final FunctionalIterable<QueryRunner<T>> queryRunners;
+    // This is to handle cases where multiple datasources may have the same SegmentDescriptor for an interval.
```

Comment:
I don't think this is going to work in all cases. Consider the case where two datasources that are being unioned both have a segment with version V1, interval 2000/P1D, and partition 0. But let's say the second datasource _also_ has one with version V2. In this case V1 should be overshadowed, but with this logic, it won't be. Both V1 and V2 will be processed for the second datasource.

I think to properly push this down to historicals, we'll need to change MultipleSpecificSegmentSpec to associate datasources with segment descriptors.

Reply:
yeah I was trying to avoid including datasources with segment descriptors as it would increase the size of the json payload transferred from the broker.